src/PosthogAnalytics.ts

```diff
@@ -0,0 +1,343 @@
+import posthog, { PostHog } from 'posthog-js';
+import PlatformPeg from './PlatformPeg';
+import SdkConfig from './SdkConfig';
+import SettingsStore from './settings/SettingsStore';
+
+/* Posthog analytics tracking.
+ *
+ * Anonymity behaviour is as follows:
+ *
+ * - If Posthog isn't configured in `config.json`, events are not sent.
+ * - If [Do Not Track](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/doNotTrack) is
+ *   enabled, events are not sent (this detection is built into posthog and turned on via the
+ *   `respect_dnt` flag being passed to `posthog.init`).
+ * - If the `feature_pseudonymousAnalyticsOptIn` labs flag is `true`, track pseudonomously, i.e.
+ *   hash all matrix identifiers in tracking events (user IDs, room IDs etc) using SHA-256.
+ * - Otherwise, if the existing `analyticsOptIn` flag is `true`, or not present (i.e. prior to
+ *   logging in), track anonymously, i.e. redact all matrix identifiers in tracking events.
+ * - If both flags are false, events are not sent.
+*/
+
+interface IEvent {
+    // The event name that will be used by PostHog.
+    // TODO: standard format (camel case? snake? UpperCase?)
+    eventName: string;
+
+    // The properties of the event that will be stored in PostHog.
+    properties: {};
```

Comment:
this should have a type, e.g `Record<string, any>;` rather than a blank interface

Reply:
also it seems like it is optional, given `IWelcomeScreenLoad` doesn't have one so maybe `?:`

Reply:
The idea is that any "concrete" (i.e. actually used to pass a concrete event) interface extending IEvent should override `properties` to provide a concrete definition of types. 

I'm not sure of a way in TypeScript to enforce that - you can't declare these things as abstract, afaik. 

Declaring `properties` as something specific like `Record<string, any>` prevents overriding it with something else, e.g. `properties: { foo: string }` - compiler says no. 

The code for IWelcomeScreen was a mistake - I got distracted while working out how to declare "no properties" which is now fixed by declaring `properties: Record<any, never>;`, and left my test wrong invocation in place (`{foo: bar}`) 🙈 .

Reply:
Another approach could have been generics, but this is fine

Reply:
definitely keen to understand that approach just out of interest

Reply:
```
interface IEvent<P extends {} | void> {
...
    properties?: P;
...
}

though I don't think it has any gains above this tbh